### PR TITLE
[#264] user header cxx api

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

Relates to #264

### CODE REVIEW

File: iceoryx2-ffi/ffi/src/api/sample.rs [hunk 1]
```diff
@@ -116,6 +116,30 @@ pub unsafe extern "C" fn iox2_cast_sample_ref_h(handle: iox2_sample_h) -> iox2_s
...
+
+    let sample = &mut *sample_handle.as_type();
+
+    let header = match sample.service_type {
+        iox2_service_type_e::IPC => sample.value.as_mut().ipc.user_header(),
+        iox2_service_type_e::LOCAL => sample.value.as_mut().local.user_header(),
+    };
+
+    *header_ptr = (header as *const UserHeaderFfi).cast();
+}
```
Comment (elBoberido) [inline iceoryx2-ffi/ffi/src/api/sample.rs:141]: Would it make sense to set `*header_ptr = std::ptr::null();` when no user header is configured, i.e. the size is 0?

File: iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp [hunk 2]
```diff
@@ -314,4 +320,44 @@ TYPED_TEST(ServicePublishSubscribeTest, publisher_applies_unable_to_deliver_stra
...
+TYPED_TEST(ServicePublishSubscribeTest, send_receive_with_user_header_works) {
+    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;
+
+    const auto* name_value = "I am floating through the galaxy of my brain. Oh the colors!";
+    const auto service_name = ServiceName::create(name_value).expect("");
+
+    auto node = NodeBuilder().create<SERVICE_TYPE>().expect("");
+    auto service_pub = node.service_builder(service_name)
+                           .template publish_subscribe<uint64_t>()
+                           .template user_header<TestHeader>()
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp:333]:
```suggestion
                           .publish_subscribe<uint64_t>()
                           .user_header<TestHeader>()
```

File: iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp [hunk 3]
```diff
@@ -314,4 +320,44 @@ TYPED_TEST(ServicePublishSubscribeTest, publisher_applies_unable_to_deliver_stra
...
+                           .create()
+                           .expect("");
+    auto service_sub = node.service_builder(service_name)
+                           .template publish_subscribe<uint64_t>()
+                           .template user_header<TestHeader>()
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp:338]:
```suggestion
                           .publish_subscribe<uint64_t>()
                           .user_header<TestHeader>()
```

**Review by elBoberido — COMMENTED** [review]
Just the usual nitpicking

File: iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp:333]: does not compile.

File: iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp (hunk 3 above)
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp:338]: does not compile.

File: iceoryx2-ffi/ffi/src/api/sample.rs (hunk 1 above)
Comment (elfenpiff) [reply iceoryx2-ffi/ffi/src/api/sample.rs:141]: Yes, but sadly I do not have this information at this level and must completely trust the user. On Rust and C++ the type system handles that for me but not here.

File: iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx2-ffi/cxx/tests/src/service_publish_subscribe_tests.cpp:333]: I'm disappointed you gave up so easily ... just write a new C++ compiler ;)

**Review by elBoberido — APPROVED** [review]

**Pull request merged**